moesif-servlet/src/main/java/com/moesif/servlet/MoesifFilter.java

```diff
@@ -427,6 +444,143 @@ private EventResponseModel getEventResponseModel(LoggingHttpServletResponseWrapp
+   */
+  public void initBatchProcessorAndStartJobs() {
+
+    // Create event batch processor for queueing and batching the events.
+    this.batchProcessor = new BatchProcessor(this.moesifApi, this.config, this.debug);
+
+    // Initialize the timer tasks - Create scheduled jobs
+    this.scheduleAppConfigJob();
+    this.scheduleBatchEventsJob();
+  }
+
+  /***
+   * Method to stop the scheduled jobs and drain the batch queue.
+   * Batch event processor sends the leftover events present in queue
+   * before stopping the scheduled jobs.
+   */
+  public void drainQueueAndStopJobs() {
+
+    // Cleanup/transfer the leftover queue events, if any before destroying the filter.
+    this.batchProcessor.run();  // it's ok to run in main thread on destroy.
+
+    // Stop the scheduled jobs
+    try {
+      if (debug) {
+        logger.info("Stopping scheduled jobs.");
+      }
+      this.resetJobTimer(this.updateConfigTimer);
+      this.resetJobTimer(this.sendBatchEventTimer);
+    } catch (Exception e) {
+      // ignore the error.
+    }
+  }
+
+  /**
+   * Method to create scheduled job for updating config periodically.
+   */
+  private void scheduleAppConfigJob() {
+    // Make sure there is none before creating the timer
+    this.resetJobTimer(this.updateConfigTimer);
+
+    this.updateConfigTimer = new Timer("moesif_update_config_job");
+    updateConfigTimer.schedule(new TimerTask() {
+      @Override
+      public void run() {
+        // get and update config.
+        getAndUpdateAppConfig();
+      }
+    }, 0, (long) this.batchProcessor.getUpdateConfigTimeInMin() * 60 * 1000);
+  }
+
+  /**
+   * Method to create scheduled job for sending batch events periodically.
+   */
+  private void scheduleBatchEventsJob() {
+    // Make sure there is none before creating the timer
+    this.resetJobTimer(this.sendBatchEventTimer);
+
+    this.sendBatchEventTimer = new Timer("moesif_events_batch_job");
+    sendBatchEventTimer.schedule(
+            this.batchProcessor,
+            0,
+            (long) this.batchProcessor.getBatchMaxTimeInSec() * 1000
+    );
+  }
+
+  /**
+   * Method to reset scheduled job timer.
+   * @param timer : Timer
+   */
+  private void resetJobTimer(Timer timer) {
+    if (timer != null) {
+      timer.cancel();
+      timer.purge();
+      timer = null;
+    }
+  }
+
+  private void rescheduleSendEventsJobIfNeeded() {
+
+    // Check if batchJob is already running then return
+    if (this.batchProcessor.isJobRunning()) {
+      if (debug) {
+        String msg = String.format("Send event job is in-progress.");
+        logger.info(msg);
+      }
+      return;
+    }
+
+    // Check if we need to reschedule job to send batch events
+    // if the last job runtime is more than 5 minutes.
+    final long MAX_TARDINESS_SEND_EVENT_JOB = this.config.batchMaxTimeInSec * 60;      // in seconds
+    final long diff = new Date().getTime() - this.batchProcessor.scheduledExecutionTime();
```

Comment:
this should be safe for possible clock adjustments such as daylight savings time change, i think?

Reply:
Good question @CodeYouMust I assumed server runs in UTC zone and locale is handled by client to avoid TZ pitfalls. Please correct me if I am wrong.

Reply:
@praves77 As far as server tz, since this is public sdk, we cant assume UTC. but I think since this is long time it should be tz agnostic, just rusty on it, so should confirm in all places where we check time. if system time changes as a result of clock sync change (eg ntp sync), then i dunno, but that should be very unlikely that it changes more than several seconds. Just maybe defensively ensure we dont have `long diff` beyond certain positive number range.